flytekit/extras/tasks/shell.py

```diff
@@ -213,21 +238,23 @@ def execute(self, **kwargs) -> typing.Any:
             print(gen_script)
             print("\n==============================================\n")
 
-        try:
-            if platform.system() == "Windows" and os.environ.get("ComSpec") is None:
-                # https://github.com/python/cpython/issues/101283
-                os.environ["ComSpec"] = "C:\\Windows\\System32\\cmd.exe"
-            subprocess.check_call(gen_script, shell=True)
-        except subprocess.CalledProcessError as e:
+        if platform.system() == "Windows" and os.environ.get("ComSpec") is None:
+            # https://github.com/python/cpython/issues/101283
+            os.environ["ComSpec"] = "C:\\Windows\\System32\\cmd.exe"
+
+        returncode, stdout, stderr = _run_script(gen_script)
+        if returncode != 0:
             files = os.listdir(".")
             fstr = "\n-".join(files)
-            logger.error(
-                f"Failed to Execute Script, return-code {e.returncode} \n"
-                f"StdErr: {e.stderr}\n"
-                f"StdOut: {e.stdout}\n"
-                f" Current directory contents: .\n-{fstr}"
+            error = (
+                f"Failed to Execute Script, return-code {returncode} \n"
+                f"Current directory contents: .\n-{fstr}\n"
+                f"StdOut: {stdout}\n"
+                f"StdErr: {stderr}\n"
             )
-            raise
+            logger.error(error)
+            # raise FlyteRecoverableException so that it's classified as user error and will be retried
+            raise FlyteRecoverableException(error)
```

Comment:
I think not every user error should be a recoverable one 🤔 

Take a look at [how `PythonFunctionTask` separates user vs system error scopes](https://github.com/flyteorg/flytekit/blob/982ffe6097dd6700d084ff4da276fe4315927e77/flytekit/core/python_function_task.py#L164):

```python
return exception_scopes.user_entry_point(self._task_function)(**kwargs)
```

Reply:
Right, of course, makes sense :+1:
This change is not really backwards compatible though, I wonder whether this is a dealbreaker 🤔

Reply:
Yes, there is undoubtedly a different behavior change when a shell task throws an error.

Before: Error in shell task is considered system error thus retried N number of times, where N is [3 by default](https://github.com/flyteorg/flytepropeller/blob/6a14e7fbffe89786fb1d8cde22715f93c2f3aff5/pkg/controller/config/config.go#L97C1-L98C1). 

After: Error in shell task is considered user recoverable error. By default, it's not retried and will only be retried if users specify retry in the task's metadata. 

 However, the shell task didn't respect the retry strategy set by users, and I think it can be considered as bug.

Reply:
@wild-endeavor your opinion will be very helpful here.

Reply:
+1, I think shell task should throw user recoverable error